serde_with/Cargo.toml

```
@@ -65,6 +65,7 @@ time_0_3 = ["dep:time_0_3"]
 base64 = {version = "0.21.0", optional = true, default-features = false}
 chrono_0_4 = {package = "chrono", version = "0.4.20", optional = true, default-features = false, features = ["serde"]}
 doc-comment = {version = "0.3.3", optional = true}
+hashbrown = {version = "0.14.0", optional = true, features = ["serde"]}
```

The crate name should be versioned. If in the future hashbrown v0.15 comes out, it should be possible to support it, without extra troubles. By adding the version into the imported name it is much simpler to support multiple versions of the same dependency.

None of the default features seem important, so I would disable them all. All implementations in serde_with should be independent of the hasher if possible, thus no need for `ahash`. `inline-more` is only a performance optimization and can be enabled by downstream crates. `allocator-api2` seems only relevant for raw types.

```suggestion
hashbrown_0_14 = {version = "0.14.0", optional = true, default-features = false, features = ["serde"]}
```

You should add a matching feature above too. Hashbrown seems to require the `alloc` crate, so alloc support should be enabled whenever using hashbrown too.

```toml
hashbrown_0_14 = ["dep:hashbrown_0_14", "alloc"]
``````

---

If we drop the `ahash` dependency, I believe we must declare our own default hasher. Some questions I have that you may already know the answer to:

- Should we use ahash as default.
- Should we use the default hasher from std::collections::HashMap.
  - If we choose this one, is that hasher available in #[no_std]?
- Does our selection of default hasher affect downstream users who would expect ahash as the default, I would assume not and that the macros would run on their namespace and thus use their hashers?

---

There should be no need to declare a default hasher. Rather, the code should be generic to work with most/any kind of hasher and should not assume anything about it. At least for serialization, this works flawlessly. Iterating over a HashMap does not require hashing anything, so no bounds need to be placed. For deserializing hashing is needed. But the HashMap only really cares that it can get a `Hasher`, which is what the `BuildHasher` trait gives. This leaves the question of how to create the `BuildHasher`. The simplest solution is `Default`. That will not always work, but most implement `Default`. That is why the deserialization code has `S: BuildHasher + Default` as the bounds.

This might leave the tests without a defined hasher. But all tests only run on targets with std so the `std::collections::hash_map::DefaultHasher` type can be used there.

Any user of hashbrown and serde_with can still specify the hasher they want and that can be ahash.

---

Looking at `hashbrown` it appears disabling `ahash` leaves `HashMap` and `HashSet` without a default hash builder:
https://github.com/rust-lang/hashbrown/blob/b2190e22e82b0b3fbae14f508bf5fc08a977df80/src/map.rs#L17

I also am unaware of how we can work around this short of a PR to `hashbrown` allowing `RandomState` to be used as a default hasher (but that would break the additive nature of feature flags so unlikely they would accept it).

As such, it stands to reason we cannot compile the hashbrown_0_14 without the `ahash` flag, is my assessment correct?

---

No code in serde_with should make any assumption about what the default hash builder is. All implementations should be generic over that parameter. As such, removing that should not cause any problems and if it does, then the implementations need to be fixed.

The one exception to that are the tests. Here you do need concrete instances of hashmaps, which means you need a concrete hasher too. The tests only run on targets with std available. The dev-dependencies contain [`fnv`](https://docs.rs/fnv) too. Between `std::collections::hash_map::DefaultHasher` and `fnv::FnvHasher` there should be no need for a third hasher. If the implementations work with both hashers, the implementations are sufficiently generic to support ahash too.

Maybe I am missing something here. Buy my plan would be to remove the feature. See if the implementations in serde_with still compile. In the tests it might be necessary to specify either of the above hasher manually.

---

If you remove the feature, serde_with no longer compiles. My belief is its because the fallback default hasher is bogus:
https://github.com/rust-lang/hashbrown/blob/b2190e22e82b0b3fbae14f508bf5fc08a977df80/src/map.rs#L17

Perhaps I've missdiagnosed, are you able to checkout and build without ahash?